Fix stale initial sync/mining in devnet

codablock opened the pull request:
With the current rules, we always assume that there is some peer somewhere in the universe that has more recent blocks then we have when the node is started. This is not fully true for devnet, as even the mining nodes start with an empty chain.

See commit descriptions for details on fixes.

gladcow commented:
utACK

UdjinM6 commented:
First blocks have ultra-low difficulty, so you don't need `getblocktemplate`, all you need to mine is `generate`. You can also easily run 2 devnet nodes connected to each other on a local machine if you want to use `getblocktemplate` later for some reason. This 2 simple things work, I just tested it, so I'd rather use them instead of adding yet another `if dev than this` code branch.

nACK

vikramdurai commented:
I think this should be closed...?

UdjinM6 commented:
@vikramdurai Yep, will close for now.
@codablock feel free to re-open if I'm missing smth

codablock commented:
@UdjinM6 The solution you proposed only works if the nodes connect to the first node after `generate` was called. If they connect before the call to `generate`, they'll never be able to sync the initial block because delaying of GETHEADERS kicks in, which in turn will never send the delayed GETHEADERS because the IBD mode is never left (no blocks arrive).

These are the 2 scenarios:
# Scenario 1 (the one you probably tested)
1. Node1 starts up
2. You call `generate 1`
3. Node1 leaves IBD mode because we now have a recent chain tip
4. Node2 starts up and connects to Node1
5. Node2 is in IBD and requests headers from Node1
6. Node1 sends HEADERS to Node2
7. Node2 requests blocks from Node1
8. All good now. Both can use `generate` and `getblocktemplate` now

# Scenario 2
1. Node1 starts up
2. Node2 starts up and connects to Node1
3. Both nodes are in IBD
4. Both nodes send GETHEADERS to the other one
5. Both nodes ignore the GETHEADERS request as they are in IBD
6. You call `generate` on Node1
7. Node1 announces new header to Node2 via INV
8. Node2 receives INV and puts the block hash into the delayed headers list as it is in IBD mode
9. As Node2 never leaves IBD, it'll never send the delayed GETHEADERS
10. Both nodes are stalled now
11. If Node2 is restarted now, it'll sync correctly

It'll be messy to workaround this in a automated setup. It'll require a dedicated node for seeding the initial block which has to be started with `generate` called before the other nodes. Also, this will be hard to debug and understand for anyone who doesn't fully understand the details of IBD.

If you want, I can change the if-devnet logic to use a chainparams boolean instead, but I'm not sure if it helps much in regard to maintainability. Initially I was thinking about setting delayGetHeadersTime to 0 for devnet (as in regtest) but that would disable the delaying logic completely on devnet, which I'd like to avoid so that devnet is as much as testnet/mainnet as possible.

UdjinM6 commented:
Tested Scenario 2 and it indeed fails. Still NACK for bc09a7e but maybe 53e8a23 is a good one. Will play around this a bit.

codablock commented:
Removed https://github.com/dashpay/dash/commit/bc09a7e42810fe350e3fa7e87677634675444aed now and added another commit that simply disables delayed getheaders on devnet.

codablock commented:
Big confusion on my side, re-pushed a version of this PR with only https://github.com/dashpay/dash/commit/53e8a23520ea46bc8d906fec46a9486c794b2c84
